File: bindings/julia/src/schema.jl
# schema.jl - schema validation against spec/schema/*.schema.json.
#
# A deliberately small interpreter for exactly the JSON Schema keywords the
# eight Causalontology schemas use: type, const, enum, pattern, required,
# properties, additionalProperties, items, minItems, minLength, minimum,
# maximum, oneOf, and local ref (#/defs/...).  "format" is treated as an
# annotation, as the 2020-12 draft does by default.

# kind -> schema file.  Three token kinds keep their original 1.0.0-reserved
# file names (individual/token/state); the id scheme is the whole word.
const SCHEMA_FILES = Dict{String,String}(
    "occurrent"  => "occurrent.schema.json",
    "causal_relation_object" => "causal_relation_object.schema.json",
    "continuant" => "continuant.schema.json",
    "realizable" => "realizable.schema.json",
    "stratum"    => "stratum.schema.json",
    "bridge"     => "bridge.schema.json",
    "port"       => "port.schema.json",
    "conduit"    => "conduit.schema.json",
    "quality"    => "quality.schema.json",
    "token_individual"   => "individual.schema.json",
    "token_occurrence"   => "token.schema.json",
    "state_assertion"    => "state.schema.json",
    "token_causal_claim" => "token_causal_claim.schema.json",
    "assertion"  => "assertion.schema.json",
    "enrichment" => "enrichment.schema.json",
    "retraction" => "retraction.schema.json",
    "succession" => "succession.schema.json",
)

const _schema_cache = Dict{String,JObj}()   # by filename
const _SCHEMA_BASE = "https://causalontology.org/schema/"

function _schema_dir()
    env = get(ENV, "CAUSALONTOLOGY_SPEC", nothing)
    env !== nothing && return joinpath(env, "schema")
    # src/ -> julia/ -> bindings/ -> repository root
    return normpath(joinpath(@__DIR__, "..", "..", "..", "spec", "schema"))
end

function _load_file(filename::AbstractString)
    get!(_schema_cache, filename) do
        json_parse(read(joinpath(_schema_dir(), filename), String))
    end
end

function load_schema(kind)
    haskey(SCHEMA_FILES, kind) || error("unknown kind: $(repr(kind))")
    return _load_file(SCHEMA_FILES[kind])
end

function _navigate(doc, pointer::AbstractString)
    node = doc
    for part in split(pointer, '/')
        part == "" && continue
        node = node[String(part)]
    end
    return node
end

"Resolve local and cross-file \$refs to a concrete schema node + its root."
function _resolve_ref(schema, root::JObj)
    while schema isa JObj && jhas(schema, "\$ref")
        ref = jget(schema, "\$ref")
        if startswith(ref, "#/")
            schema = _navigate(root, ref[3:end])
        elseif startswith(ref, _SCHEMA_BASE)
            rest = ref[length(_SCHEMA_BASE)+1:end]
            idx = findfirst("#/", rest)
            if idx === nothing
                filename, pointer = rest, ""
            else
                filename = rest[1:first(idx)-1]
                pointer = rest[first(idx)+2:end]
            end
            root = _load_file(filename)
            schema = pointer == "" ? root : _navigate(root, pointer)
        else
            error("unsupported \$ref: $ref")
        end
    end
    return schema, root
end

function _type_ok(value, t)
    t == "object"  && return value isa JObj
    t == "array"   && return value isa AbstractVector
    t == "string"  && return value isa AbstractString
    t == "boolean" && return value isa Bool
    # "number"/"integer": never Bool (mirrors the Python check).  Julia parses
    # every integer literal to Int64, so "integer" accepts an Int64 value.
    t == "number"  && return value isa Union{Int64,Float64}
    t == "integer" && return value isa Int64
    return false
end

function _schema_check(value, schema, root::JObj, path::String,
                       errors::Vector{String})
    schema, root = _resolve_ref(schema, root)

    oneof = jget(schema, "oneOf")
    if oneof !== nothing
        passing = 0
        for sub in oneof
            suberrs = String[]
            _schema_check(value, sub, root, path, suberrs)
            isempty(suberrs) && (passing += 1)
        end
        passing != 1 && push!(errors,
            "$path: matches $passing of the oneOf branches (need exactly 1)")
        return
    end

    t = jget(schema, "type")
    if t !== nothing
        if !_type_ok(value, t)
            push!(errors, "$path: expected $t")
            return
        end
    end

    if jhas(schema, "const") && value != jget(schema, "const")
        push!(errors, "$path: must equal $(repr(jget(schema, "const")))")
    end
    if jhas(schema, "enum") && !any(e -> e == value, jget(schema, "enum"))
        push!(errors, "$path: $(repr(value)) not in enumeration")
    end
    if jhas(schema, "pattern") && value isa AbstractString
        if !occursin(Regex(jget(schema, "pattern")), value)
            push!(errors, "$path: $(repr(value)) does not match " *
                          jget(schema, "pattern"))
        end
    end
    if jhas(schema, "minLength") && value isa AbstractString
        length(value) < jget(schema, "minLength") &&
            push!(errors, "$path: shorter than minLength")
    end
    if jhas(schema, "minimum") && value isa Union{Int64,Float64}
        value < jget(schema, "minimum") &&
            push!(errors, "$path: below minimum $(jget(schema, "minimum"))")
    end
    if jhas(schema, "maximum") && value isa Union{Int64,Float64}
        value > jget(schema, "maximum") &&
            push!(errors, "$path: above maximum $(jget(schema, "maximum"))")
    end

    if value isa AbstractVector
        if jhas(schema, "minItems") && length(value) < jget(schema, "minItems")
            push!(errors,
                  "$path: fewer than $(jget(schema, "minItems")) items")
        end
        if jhas(schema, "items")
            for (i, item) in enumerate(value)
                # the error path keeps 0-based indices, as the reference does
                _schema_check(item, jget(schema, "items"), root,
                              "$path[$(i - 1)]", errors)
            end
        end
    end

    if value isa JObj
        props = jget(schema, "properties", JObj())
        for req in jget(schema, "required", Any[])
            jhas(value, req) ||
                push!(errors, "$path: required property '$req' missing")
        end
        if jget(schema, "additionalProperties") === false
            for k in jkeys(value)
                jhas(props, k) ||
                    push!(errors, "$path: additional property '$k'")
            end
        end
        for (k, sub) in props.pairs
            if jhas(value, k)
                _schema_check(value[k], sub, root, "$path.$k", errors)
            end
        end
    end
end

"(ok, reasons) - structural validity against the kind's JSON Schema."
function validate_schema(obj::JObj, kind=nothing)
    k = kind === nothing ? infer_kind(obj) : kind
    root = load_schema(k)
    errors = String[]
    _schema_check(obj, root, root, "\$", errors)
    return isempty(errors), errors
end

File: bindings/dart/lib/schema.dart
/// Schema validation against spec/schema/*.schema.json.
///
/// A deliberately small interpreter for exactly the JSON Schema keywords the
/// seventeen Causalontology schemas use: type, const, enum, pattern, required,
/// properties, additionalProperties, items, minItems, minLength, minimum,
/// maximum, oneOf, local $ref (#/$defs/...), and cross-file $ref to a sibling
/// schema (https://causalontology.org/schema/<file>.schema.json#/...).
/// "format" is treated as an annotation, as the 2020-12 draft does by default.
library;

import 'dart:convert';
import 'dart:io';

import 'canonical.dart';

/// kind -> schema file. Three token kinds keep their original 1.0.0-reserved
/// file names (individual/token/state); the id scheme is the whole word.
const Map<String, String> schemaFiles = {
  'occurrent': 'occurrent.schema.json',
  'causal_relation_object': 'causal_relation_object.schema.json',
  'continuant': 'continuant.schema.json',
  'realizable': 'realizable.schema.json',
  'stratum': 'stratum.schema.json',
  'bridge': 'bridge.schema.json',
  'port': 'port.schema.json',
  'conduit': 'conduit.schema.json',
  'quality': 'quality.schema.json',
  'token_individual': 'individual.schema.json',
  'token_occurrence': 'token.schema.json',
  'state_assertion': 'state.schema.json',
  'token_causal_claim': 'token_causal_claim.schema.json',
  'assertion': 'assertion.schema.json',
  'enrichment': 'enrichment.schema.json',
  'retraction': 'retraction.schema.json',
  'succession': 'succession.schema.json',
};

const String _base = 'https://causalontology.org/schema/';

/// Cache keyed by schema-file name (cross-file $ref shares this cache).
final Map<String, Map<String, dynamic>> _cache = {};

/// Locate the repository's spec/schema directory: the CAUSALONTOLOGY_SPEC
/// environment variable (naming the spec/ directory) wins; otherwise walk
/// up from this library's own location, then from the working directory,
/// until a spec/schema directory is found.
Directory schemaDir() {
  final env = Platform.environment['CAUSALONTOLOGY_SPEC'];
  if (env != null && env.isNotEmpty) {
    return Directory('$env${Platform.pathSeparator}schema');
  }
  final starts = <Directory>[
    File.fromUri(Platform.script).parent,
    Directory.current,
  ];
  for (final start in starts) {
    var dir = start.absolute;
    while (true) {
      final candidate = Directory(
          '${dir.path}${Platform.pathSeparator}spec${Platform.pathSeparator}schema');
      if (candidate.existsSync()) return candidate;
      final parent = dir.parent;
      if (parent.path == dir.path) break;
      dir = parent;
    }
  }
  throw StateError('cannot locate spec/schema; set CAUSALONTOLOGY_SPEC');
}

Map<String, dynamic> _loadFile(String filename) {
  return _cache.putIfAbsent(filename, () {
    final path = '${schemaDir().path}${Platform.pathSeparator}$filename';
    return jsonDecode(File(path).readAsStringSync()) as Map<String, dynamic>;
  });
}

Map<String, dynamic> loadSchema(String kind) {
  if (!schemaFiles.containsKey(kind)) {
    throw ArgumentError('unknown kind: $kind');
  }
  return _loadFile(schemaFiles[kind]!);
}

Map<String, dynamic> _navigate(Map<String, dynamic> doc, String pointer) {
  dynamic node = doc;
  for (final part in pointer.split('/')) {
    if (part.isEmpty) continue;
    node = (node as Map)[part];
  }
  return (node as Map).cast<String, dynamic>();
}

/// Resolve local and cross-file $refs to a concrete schema node + its root.
(Map<String, dynamic>, Map<String, dynamic>) _resolve(
    Map<String, dynamic> schema, Map<String, dynamic> root) {
  var node = schema;
  var r = root;
  while (node.containsKey(r'$ref')) {
    final ref = node[r'$ref'] as String;
    if (ref.startsWith('#/')) {
      node = _navigate(r, ref.substring(2));
    } else if (ref.startsWith(_base)) {
      final rest = ref.substring(_base.length);
      final hashIdx = rest.indexOf('#/');
      final filename = hashIdx < 0 ? rest : rest.substring(0, hashIdx);
      final pointer = hashIdx < 0 ? '' : rest.substring(hashIdx + 2);
      r = _loadFile(filename);
      node = pointer.isEmpty ? r : _navigate(r, pointer);
    } else {
      throw ArgumentError('unsupported \$ref: $ref');
    }
  }
  return (node, r);
}

/// Deep structural equality over the JSON model (maps, lists, scalars).
bool deepEquals(Object? a, Object? b) {
  if (a is Map && b is Map) {
    if (a.length != b.length) return false;
    for (final key in a.keys) {
      if (!b.containsKey(key) || !deepEquals(a[key], b[key])) return false;
    }
    return true;
  }
  if (a is List && b is List) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (!deepEquals(a[i], b[i])) return false;
    }
    return true;
  }
  if (a is num && b is num && a is! bool && b is! bool) {
    return a == b; // 1 == 1.0 holds in Dart, matching Python's ==
  }
  return a == b;
}

bool _typeMatches(Object? value, String t) {
  switch (t) {
    case 'object':
      return value is Map;
    case 'array':
      return value is List;
    case 'string':
      return value is String;
    case 'number':
      return value is num && value is! bool;
    case 'integer':
      return value is int;
    case 'boolean':
      return value is bool;
    default:
      throw ArgumentError('unknown schema type: $t');
  }
}

void _check(Object? value, Map<String, dynamic> schemaNode,
    Map<String, dynamic> rootIn, String path, List<String> errors) {
  final (schema, root) = _resolve(schemaNode, rootIn);

  if (schema.containsKey('oneOf')) {
    var passing = 0;
    for (final sub in schema['oneOf'] as List) {
      final suberrs = <String>[];
      _check(value, (sub as Map).cast<String, dynamic>(), root, path, suberrs);
      if (suberrs.isEmpty) passing++;
    }
    if (passing != 1) {
      errors.add('$path: matches $passing of the oneOf branches '
          '(need exactly 1)');
    }
    return;
  }

  final t = schema['type'] as String?;
  if (t != null) {
    if (!_typeMatches(value, t)) {
      errors.add('$path: expected $t');
      return;
    }
  }

  if (schema.containsKey('const') && !deepEquals(value, schema['const'])) {
    errors.add('$path: must equal ${schema['const']}');
  }
  if (schema.containsKey('enum') &&
      !(schema['enum'] as List).any((e) => deepEquals(value, e))) {
    errors.add('$path: $value not in enumeration');
  }
  if (schema.containsKey('pattern') && value is String) {
    if (!RegExp(schema['pattern'] as String).hasMatch(value)) {
      errors.add("$path: '$value' does not match ${schema['pattern']}");
    }
  }
  if (schema.containsKey('minLength') && value is String) {
    if (value.length < (schema['minLength'] as num)) {
      errors.add('$path: shorter than minLength');
    }
  }
  if (schema.containsKey('minimum') && value is num && value is! bool) {
    if (value < (schema['minimum'] as num)) {
      errors.add('$path: below minimum ${schema['minimum']}');
    }
  }
  if (schema.containsKey('maximum') && value is num && value is! bool) {
    if (value > (schema['maximum'] as num)) {
      errors.add('$path: above maximum ${schema['maximum']}');
    }
  }

  if (value is List) {
    if (schema.containsKey('minItems') &&
        value.length < (schema['minItems'] as num)) {
      errors.add('$path: fewer than ${schema['minItems']} items');
    }
    if (schema.containsKey('items')) {
      for (var i = 0; i < value.length; i++) {
        _check(value[i], (schema['items'] as Map).cast<String, dynamic>(),
            root, '$path[$i]', errors);
      }
    }
  }

  if (value is Map) {
    final props = (schema['properties'] as Map?)?.cast<String, dynamic>() ??
        <String, dynamic>{};
    for (final req in (schema['required'] as List?) ?? const []) {
      if (!value.containsKey(req)) {
        errors.add("$path: required property '$req' missing");
      }
    }
    if (schema['additionalProperties'] == false) {
      for (final key in value.keys) {
        if (!props.containsKey(key)) {
          errors.add("$path: additional property '$key'");
        }
      }
    }
    for (final entry in props.entries) {
      if (value.containsKey(entry.key)) {
        _check(value[entry.key], (entry.value as Map).cast<String, dynamic>(),
            root, '$path.${entry.key}', errors);
      }
    }
  }
}

/// (ok, reasons) - structural validity against the kind's JSON Schema.
(bool, List<String>) validateSchema(Map<String, dynamic> obj, [String? kind]) {
  final k = kind ?? inferKind(obj);
  final root = loadSchema(k);
  final errors = <String>[];
  _check(obj, root, root, r'$', errors);
  return (errors.isEmpty, errors);
}
